buildSrc/src/main/groovy/net/minecraftforge/gradleutils/tasks/ExtractTeamCityProjectConfigurationTask.java

```diff
@@ -0,0 +1,242 @@
+ * USA
+ */
+
+package net.minecraftforge.gradleutils.tasks;
+
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.Status;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.*;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Objects;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+/**
+ * Task to extract the TeamCity project configuration from the included template zpi file and the github remote information.
+ */
+public abstract class ExtractTeamCityProjectConfigurationTask extends DefaultTask
+{
+
+    public ExtractTeamCityProjectConfigurationTask()
+    {
+        getDestination().convention(getProject().getRootProject().getLayout().getProjectDirectory().dir(getProject().provider(() -> "./")));
+        getRequiresCleanWorkspace().convention(true);
+
+        setGroup("publishing");
+        setDescription("Creates (or recreates) a default TeamCity project configuration directory for use with the MinecraftForge TeamCity server.");
+    }
+
+    @InputDirectory
+    @PathSensitive(PathSensitivity.NONE)
+    public abstract DirectoryProperty getDestination();
+
+    @Input
+    public abstract Property<Boolean> getRequiresCleanWorkspace();
+
+    @TaskAction
+    public void run() throws Exception
+    {
+        //Get the destination directory (by default the current root project directory)
+        final File destDir = getDestination().getAsFile().get();
+        //Grab the target directory, to check if it exists.
+        final File teamcityDir = new File(destDir, ".teamcity");
+
+        //If requested validate that the repository is clean.
+        if (getRequiresCleanWorkspace().get())
+        {
+            //Check it.
+            CheckForCleanWorkspace(destDir);
+        }
+
+        //Export the zip file from our resources.
+        String fileZip = ExportResource();
+
+        //Check if the directory exists, if so then delete it.
+        if (teamcityDir.exists())
+        {
+            //Try to delete it.
+            if (!teamcityDir.delete())
+            {
+                //Something went really wrong....
+                throw new IllegalStateException("Could not delete the existing .teamcity project directory!");
+            }
+        }
+
+        //Extract the zip from our runtime file.
+        ExtractTeamCityZip(fileZip, destDir);
+        //Replace the default project ids, with ours.
+        ReplaceTeamCityTestProjectIds(destDir);
+    }
+
+    /**
+     * Extracts the .teamcity.zip file to our target directory.
+     *
+     * @param fileZip The teamcity zip file.
+     * @param destDir The target directory (generally the project directory), where the .teamcity.zip file will be extracted.
+     */
+    private static void ExtractTeamCityZip(final String fileZip, final File destDir) throws Exception
```

Comment:
Why all the weird method casing in this class?

Reply:
What do you mean weird method casing?

It is a static method, so `ExtractTeamCityZip` seems accurate....

Reply:
I presume what he means is the use of PascalCase for the method names rather than the Java convention of lowerCamelCase.

Reply:
You are right.
Java does that even for static methods.....

I will fix this tomorrow morning